src/gt4py/next/program_processors/runners/dace_iterator/__init__.py

```diff
@@ -307,3 +313,89 @@ def build_sdfg_from_itir(
+        params = {str(p.id): p.type for p in self.itir.params}
+        fields = {str(p.id): p.type for p in self.itir.params if hasattr(p.type, "dims")}
+        arg_types = list(params.values())
+
+        # Do this because DaCe converts the offset_provider to an OrderedDict with StringLiteral keys
+        offset_provider = {str(k): v for k, v in kwargs.get("offset_provider", {}).items()}
+        self.sdfg_convertible["offset_provider"] = offset_provider
+
+        sdfg = self.backend.executor.otf_workflow.step.translation.generate_sdfg(  # type: ignore[union-attr]
+            self.itir,
+            arg_types,
+            offset_provider=offset_provider,
+            column_axis=kwargs.get("column_axis", None),
+        )
+
+        sdfg.arg_names.extend(self.__sdfg_signature__()[0])
+        sdfg.arg_names.extend(list(self.__sdfg_closure__().keys()))
+
+        # TODO(kotsaloscv): Keep halo exchange related metadata (for horizontal dimension) here?
+        # Could possibly be computed directly from the SDFG.
+
+        input_fields = [
+            str(inpt.id)
+            for closure in self.itir.closures
+            for inpt in closure.inputs
+            if str(inpt.id) in sdfg.arrays
+        ]
+        sdfg.GT4Py_Program_input_fields = {
+            inpt: dim
+            for inpt in input_fields
+            for dim in fields[inpt].dims  # type: ignore[union-attr]
+            if dim.kind == common.DimensionKind.HORIZONTAL
+        }
+
+        output_fields = []
+        for closure in self.itir.closures:
+            output = closure.output
+            if isinstance(output, itir.SymRef):
+                if str(output.id) in sdfg.arrays:
+                    output_fields.append(str(output.id))
+            else:
+                for arg in output.args:
+                    if str(arg.id) in sdfg.arrays:  # type: ignore[attr-defined]
+                        output_fields.append(str(arg.id))  # type: ignore[attr-defined]
+        sdfg.GT4Py_Program_output_fields = {
+            output: dim
+            for output in output_fields
+            for dim in fields[output].dims  # type: ignore[union-attr]
+            if dim.kind == common.DimensionKind.HORIZONTAL
+        }
+
+        sdfg.offset_providers_per_input_field = {}
+        itir_tmp = apply_common_transforms(self.itir)
+        for closure in itir_tmp.closures:  # type: ignore[union-attr]
+            shifts = TraceShifts.apply(closure)
+            for k, v in shifts.items():
+                if k not in sdfg.GT4Py_Program_input_fields:
+                    continue
+                sdfg.offset_providers_per_input_field.setdefault(k, []).extend(list(v))
+
+        sdfg.stencil_horizontal_start = kwargs.get("horizontal_start", None)
+        sdfg.stencil_horizontal_end = kwargs.get("horizontal_end", None)
+
+        return sdfg
+
+    def __sdfg_closure__(self, reevaluate: Optional[Dict[str, str]] = None) -> Dict[str, Any]:
```

Comment:
What is the relation between the calls to `__sdfg_closure__()` and `__sdfg__()` from dace at run-time? I'm asking because if `__sdfg_closure__` is not called always after a call  to `__sdfg__` receiving a different offset provider, the closure could get easily out of sync easily.

Reply:
This is a nice observation. But, indeed `__sdfg_closure__ ` is always called after a call to `__sdfg__ `.